content/2019/a-test-to-attest-to/index.md

```diff
@@ -0,0 +1,1183 @@
+  the following contents:
+    * "Mar"
+    * "May"
+* Repeat for both `test/support/test_simple.vim` and
+  `test/support/test_async.vim`.
+
+Of course there are tons more cases to test, but this should do for now.
+
+At this point, if you're following along at home, satisfy yourself that the
+completion plugin works by following those test instructions. Next we'll need
+to automate them, so it makes sense to be familiar with what will happen.
+
+## Automation
+
+Finally, this is the part where we answer the question "How can I test my Vim
+plugin with pure Vim?". Yes, the question you were told you asked at the
+beginning of the article.
+
+Of course this is not a straightforward answer. Let's take it bit-by bit.
+
+## Automating typing with feedkeys
+
+It may not be obvious, but the best way to test the user typing some commands is
+to actually tell Vim to pretend that's exacty what happened. 
+
+There are ways to _actually_ enter text by running Vim in another terminal and
+sending keypresses to that terminal, but we're not going to cover that here.
+We're going to be using [`feedkeys()`][h-feedkeys].
+
+Taking a look at the help for `feedkeys()` can be a little overwhelming, but
+fear not, I've read it enough times to have a vague understanding of it. I've
+also [stepped through the Vim code with Vimspector][vimspector-vim]
+to understand exactly what's going on.
+
+### How Vim works
+
+As a terminal application, Vim is essentially a main loop a bit like this
+pseudocode:
+
+```c
+extern char *input_buffer;
+int input_len = 0;
+while (!finished) {
+  input_buffer[ input_len++ ] = read_character_from_standard_input();
+  expand_mappings_in_buffer( &input_buffer, &input_len )
+  if ( have_command_in_buffer( input_buffer ) ) {
+    handle_command_in_buffer( &input_buffer, &input_len );
+  }
+}
+```
+
+That is, read characters into a buffer until you know what to do, then do it.
+(Disclaimer: It's a tad more complex than that, but suffice to say
+there is a buffer of characters to be processed, which can contain commands made
+up of sequences of characters and that some sequences of characters can be
+expanded into some _other_ sequence of characters).
+
+### How feedkeys works
+
+Intuitively, `feedkeys()` pushes characters into the input buffer. This
+effectively tells Vim to execute as if the pushed characters were entered. But
+there are a few things you need to know:
+
+* To enter special characters, like `<C-x>`, you must use double quotes `"` and
+  a backslash `\` before the `<`, as in `feedkeys( "\<C-x>" )`
+* By default, `feedkeys()` treats characters like the _rhs_ of a mapping
+* By default, `feedkeys()` just puts things in the queue, but doesn't actually
+  execute them
+* By default, if you wait for keys to be executed, `feedkeys()` ends insert mode
+  (if it is active) when it returns, as if the command ended with `<Esc>`.
+* By default, `feedkeys()` expands mappings in the input (like `:map` rather
```

Comment:
It would, but I thought this explanation would be more familiar to most users. I suspect if you understood the normal/normal! distinction, this explanation probably isn't needed.